// @backstage/plugin-techdocs/dist/index.esm.js
export { u as DefaultTechDocsHome, a as DocsCardGrid, D as DocsTable, y as EmbeddedDocsRouter, r as EntityListDocsGrid, s as EntityListDocsTable, E as EntityTechdocsContent, R as Reader, x as Router, k as TechDocsClient, d as TechDocsCustomHome, e as TechDocsIndexPage, T as TechDocsPageWrapper, v as TechDocsPicker, c as TechDocsReaderLayout, g as TechDocsReaderPage, o as TechDocsReaderPageContent, n as TechDocsReaderPageHeader, p as TechDocsReaderPageSubheader, m as TechDocsReaderProvider, q as TechDocsSearch, h as TechDocsSearchResultListItem, l as TechDocsStorageClient, f as TechdocsPage, w as isTechDocsAvailable, t as plugin, j as techdocsApiRef, t as techdocsPlugin, i as techdocsStorageApiRef } from './esm/index-7df60ec3.esm.js';
import '@backstage/core-plugin-api';
import '@backstage/errors';
import 'event-source-polyfill';
import 'react';
import 'react-router-dom';
import '@backstage/plugin-techdocs-react';
import 'react-use/lib/useAsync';
import 'react-use/lib/useAsyncRetry';
import '@backstage/core-components';
import '@material-ui/core';
import '@backstage/plugin-search-react';
import '@material-ui/core/Typography';
import '@material-ui/lab';
import '@material-ui/icons/Close';
import '@backstage/integration-react';
import 'dompurify';
import '@backstage/integration';
import '@material-ui/icons/FeedbackOutlined';
import 'react-dom';
import 'git-url-parse';
import '@material-ui/icons/Menu';
import '@material-ui/core/IconButton';
import 'react-use/lib/useCopyToClipboard';
import 'react-helmet';
import '@material-ui/core/styles';
import '@material-ui/icons/Code';
import '@backstage/plugin-catalog-react';
import '@backstage/catalog-model';
import 'lodash';
import '@material-ui/icons/Settings';
import '@material-ui/icons/Share';
import '@material-ui/styles';
import '@material-ui/icons/Star';
import '@material-ui/icons/StarBorder';
